### ISSUE TITLE
Bound scopes incorrect after `Require`

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#10212]
This is from #10064, reported separately here as it is fixed by #10181.

On the following example:

`A.v`
```coq
Declare Scope foo_scope.
Declare Scope bar_scope.

Axiom T : Type.
Bind Scope foo_scope with T.
Definition def (x : T) := x.
Bind Scope bar_scope with T.
```

`B.v`
```coq
Require A.
About A.def.
```

One see that `A.def` is given argument scopes `foo_scope`, which is incorrect. It should in fact have no argument scopes since the bound scopes have not been `Import`ed. After `Import`, the argument scopes should `bar_scope`.

**Comment by herbelin** [comment]
Both instances of `Bind Scope` should indeed behave the same wrt `Import`. It is however not directly obvious to me that we want `Bind Scope` to be activated only at `Import`. Naively, I would say that it may be the case that we want non-qualified constants, but still have these constants bound to scopes (allowing e.g. notations from other modules).

The new semantics seems a priori more general since it gives the ability to selectively activate a `Bind Scope` by putting it in a submodule which is later explicitly imported, but this underlies a different way to organize libraries (on the other side reissuing individual `Bind Scope` is not necessarily more costly than typing `Import`).

Maybe I care too much and what importing semantics is adopted here does not matter a lot in practice since #10181 shows no difference (at least in input). I.e., in practice, `Bind Scope` maybe goes enough often in pair with `Notation` so that activating it or not does not matter until `Notation` are themselves imported.

I'd nevertheless be curious of more opinions, at least until we have more-refined ways for selective import.

**Comment by silene** [comment]
I tend to think that this is more of a feature than a bug. More precisely, I consider that the argument scopes are associated to the symbols, not to the modules currently imported.

**Comment by maximedenes** [comment]
> I tend to think that this is more of a feature than a bug. More precisely, I consider that the argument scopes are associated to the symbols, not to the modules currently imported.

I think then you are questioning #10064. In particular, why would you expect `def` to have scope `bar_scope` at the end of `A.v`?

**Comment by silene** [comment]
I don't. I expect it to have `foo_scope`.

**Comment by maximedenes** [comment]
> I don't. I expect it to have `foo_scope`.

Ok, so then your question is #10064. I had the same, but the standard library relies on `Bind Scope` having this dynamic behavior.

To be clear, this PR *does not* change the behavior of `Bind Scope`, it is fixing an obvious bug of argument scopes which do not behave the same depending on whether bound scopes where declared in the current module or an imported one. In particular, in the scenario of the bug, computed argument scopes and bound scopes disagree.

So the behavior of `Bind Scope` at `Import`, while interesting, looks like orthogonal to me.

Does it make it clearer, @herbelin? I agree these things are tricky, and I don't have a precise idea on what should be the ideal behavior of `Bind Scope` w.r.t. `Import` and whether we want the dynamic binding semantics. So I'm not changing either, just refactoring and fixing a bug.

**Comment by silene** [comment]
Not sure which question you think I am asking. I am just saying that the current documentation of `Bind Scope` is perfectly fine with me, and this report does not describe a bug, because it complies with the documentation.

**Comment by maximedenes** [comment]
> this report does not describe a bug, because it complies with the documentation.

I think that's incorrect. According to the documentation, `def` would have argument scope `foo_scope` at the end of `A.v`. It does not, it has argument scope `bar_scope`.

As we found out in #10064, this change was intentional but undocumented.

So there are (at least) two problems:
1. the documentation does  not reflect the current intended behavior (that's #10064)
1. the intended behavior is not correctly implemented (that's this issue)

The strategy I chose was to tackle 2, as it is not too disruptive and will make it easier to take a decision for 1 (either update the documentation or change the behavior).

**Comment by silene** [comment]
> According to the documentation, def would have argument scope foo_scope at the end of A.v. It does not, it has argument scope bar_scope.

That is the topic of report #10064, not of this one. Please do not mix both issues.

**Comment by maximedenes** [comment]
Ok, I think I see what you mean. I guess I'll fix #10064 by updating the documentation first, it will make things clearer.

**Comment by maximedenes** [comment]
So @herbelin, I'm a bit lost in the discussion of the "fix", but I'd like to know if this bug is considered valid or not. If not, let's simply close it, that's fine. Thanks!

**Comment by herbelin** [comment]
@maximedenes: Sure, the strange behavior described in this report is valid.

My question was whether the expected behavior is that `def` has no scope after a simple `Require` (based on the idea that `Bind Scope` is not activated by a simple `Require`) or has the scope `bar_scope` (based on the idea that `Arguments` has an effect on `Require` e.g. for implicit arguments, and thus that `BInd Scope` should be activated too even w/o import). Do you agree that there are these two alternatives?

In #10181 ([comment](https://github.com/coq/coq/pull/10181#issuecomment-496234223)), I proposed to try the second alternative to see how it works in practice. However, if time is missing, we can stay with your fix as it is (i.e. with the first alternative). As I said, the fix is good and up to the two other remarks (some dead code and the compat 8.8 to preserve by adding explicit `Bind Scope` in the compat file), I'm fully supporting it.

**Comment by maximedenes** [comment]
I don't like adding more effects on `Require` (as opposed to `Import`), since @ppedrot and I were hoping to reduce them even more. On the other hand, I don't have a strong opinion on the particular case of `Bind Scope`.

**Comment by herbelin** [comment]
> I don't like adding more effects on Require (as opposed to Import), since @ppedrot and I were hoping to reduce them even more.

Yes, I somehow agree on this and providing ways to finely control the effects of `Import` is indeed still missing (at lest compared e.g. to what Haskell or Agda provide).

But there are some effects which are attached to constants, such as implicit arguments, and it is unclear that you would really like implicit arguments to be activated only at import. Coercions (not imported) is probably more half-half. Scopes is somehow tied to the question of notations (which are not imported). But notations for a scope are not exclusively tied to the module where the scope is bound, and we may want notations imported from somewhere else to be activated even if the `Bind` of the corresponding type is in a non-imported module, so this is also questionable.

The general issue is about data visibility, and the binary import vs no-import alternative is anyway pretty crude here.

Back to #10181, are you ok with the proposal?
